proposals/pubsub-mq-interfaces-proposal.md

```diff
@@ -0,0 +1,144 @@
+
+However, if we look at services that provide `message-queue` / `pubsub` implementations, the line between the two concepts becomes blurred.
+
+That is, today's most used related services provide both `message-queue`, and `pubsub` capabilities. Below, is a visual example from [HiveMQ](https://www.hivemq.com/) displaying an event-driven architecture with both `message-queue`, and `pubsub` capabilities:
+
+![](./images/pubsub-interface-proposal-0.gif)
+Source: [HiveMQ's Shared Subscriptions](https://www.hivemq.com/docs/hivemq/4.9/user-guide/shared-subscriptions.html)
+
+##  3. <a name='pubsubhttpInteroptability'></a>`pubsub`/`http` Interoptability
+
+`pubsub`s are often used in conjuction with `http` to send, and receive messages from subscribed topics. However, through experimentation on the intereoperability between the two services, it is impossible to use `slight`'s `http` service to correctly manage `pubsub` subscriptions. This is because the `http` interface cannot return a handle to the `pubsub` client under which the subscription was created - Hence, you can only receive messages from a subscribed topic if the subscriptions is done in conjunction with the message polling. This is a problem because message polling, and subscribing should be two separate concerns as, if you only subscribe to a topic at the time of polling, you will miss messages that were sent to the topic before you subscribed to it.
+
+##  4. <a name='ProblemStatement'></a>Problem Statement
+
+1. As one of SpiderLightning's core goals is to stray away from mirroring infrastructure and focus on concepts, we should unify the `pubsub`, and `message-queue` interfaces into the umbrella of the `pubsub` pattern.
+1. The current `pubsub` interface is not well suited for event-driven architectures as:
+	- the `payload` type is lacking, and
+	- `http` interoptability is lacking - It should be modified to allow better subscription management.
+
+##  5. <a name='Proposal'></a>Proposal
+
+The change involves:
+- unify `pub` and `sub` resources into a single `pubsub` resource,
+- modify the `pubsub` interface to contain `message-queue` semantics,
+- change the `payload` type to `event` (following the `CloudEvents` spec),
+- remove the `subscribe` method from the `pubsub` interface to pass the responsibility of subscriptions to `slight` through information made available at compilation-time through the `slightfile`, and
+- add the `unsubscribe` method to the `pubsub` interface.
+
+###  5.1. <a name='pubsubWIT'></a>`pubsub` WIT
+
+```wit
+// A Pub/Sub Interface
+use { error, event } from types
+
+variant channel {
+	queue(string),
+	topic(string)
+}
+
+resource pubsub {
+	// creates a handle to a pubsub object
+	static open: func(name: string) -> expected<pubsub, error>
+
+	// unsubscribes from a channel
+	unsubscribe: func(location: channel) -> expected<unit, error>
```

Comment:
Do we really need this method?

Reply:
Considering subscriptions are handled by the runtime, I figured it would be a good idea to, at least, be able to unsubscribe - What are your thoughts?